ouroboros-consensus-cardano/app/db-converter.hs

```diff
@@ -0,0 +1,266 @@
+import qualified Data.ByteString.Lazy as BSL
+import           Data.Foldable
+import           Data.Functor
+import           Data.Maybe
+import           Data.Text hiding (foldl')
+import qualified Database.LMDB.Simple as LMDB
+import qualified Database.LMDB.Simple.Cursor as LMDB.Cursor
+import           DBAnalyser.Parsers
+import           Main.Utf8
+import           Options.Applicative
+import           Ouroboros.Consensus.Block
+import           Ouroboros.Consensus.Config
+import           Ouroboros.Consensus.Ledger.Basics
+import           Ouroboros.Consensus.Ledger.Extended
+import           Ouroboros.Consensus.Ledger.Tables.Utils
+import           Ouroboros.Consensus.Node.ProtocolInfo
+import           Ouroboros.Consensus.Storage.LedgerDB.Impl.Common
+import           Ouroboros.Consensus.Storage.LedgerDB.Impl.Snapshots
+import qualified Ouroboros.Consensus.Storage.LedgerDB.V1.BackingStore.Impl.LMDB as Disk
+import qualified Ouroboros.Consensus.Storage.LedgerDB.V1.BackingStore.Impl.LMDB.Bridge as LMDB.Bridge
+import           Ouroboros.Consensus.Util.CBOR
+import           Ouroboros.Consensus.Util.IOLike
+import           System.Exit
+import           System.FS.API
+import           System.FS.API.Lazy
+import           System.FS.IO
+
+data Mode
+    = Legacy
+    | Mem
+    | LMDB
+    deriving (Show, Read)
+
+data Config = Config
+    { from    :: Mode
+    -- ^ Which mode the input snapshot is in
+    , inpath  :: FsPath
+    -- ^ Path to the input snapshot
+    , to      :: Mode
+    -- ^ Which mode the output snapshot must be in
+    , outpath :: FsPath
+    -- ^ Path to the output snapshot
+    }
+
+getCommandLineConfig :: IO (Config, BlockType)
+getCommandLineConfig =
+    execParser $
+        info
+            ((,) <$> parseConfig <*> blockTypeParser <**> helper)
+            (fullDesc <> progDesc "Utility for converting snapshots to and from UTxO-HD")
+
+parseConfig :: Parser Config
+parseConfig =
+    Config
+        <$> argument
+            auto
+            ( mconcat
+                [ help "From mode (Legacy, Mem or LMDB)"
+                , metavar "MODE-IN"
+                ]
+            )
+        <*> ( fromMaybe (error "Malformed input path") . fsFromFilePath (MountPoint ".")
+                <$> strArgument
+                    ( mconcat
+                        [ help "Input dir/file"
+                        , metavar "PATH-IN"
+                        ]
+                    )
+            )
+        <*> argument
+            auto
+            ( mconcat
+                [ help "To mode (Legacy, Mem or LMDB)"
+                , metavar "MODE-OUT"
+                ]
+            )
+        <*> ( fromMaybe (error "Malformed output path") . fsFromFilePath (MountPoint ".")
+                <$> strArgument
+                    ( mconcat
+                        [ help "Output dir/file"
+                        , metavar "PATH-OUT"
+                        ]
+                    )
+            )
```

Comment:
What would happen if the user puts in an absolute path instead of a relative path? I'm unsure what would happen. We should either tell the user to only pass in only relative paths (and check that they do), or check that it's fine to use absolute paths

Reply:
I asked the user now to input relative paths, although I still don't check them. I don't know how to check them with `fs-api`

Reply:
Done